ledger/fullblock_perf_test.go

```diff
@@ -0,0 +1,331 @@
+	if err != nil {
+		return nil, nil, basics.Address{}, nil, nil, err
+	}
+
+	return l0, l1, creator, accts, aIdxs, nil
+}
+
+func BenchmarkBlockValidationMix(b *testing.B) {
+	numAssets := 10000
+	benchmarkBlockValidationMix(b, numAssets)
+}
+
+func BenchmarkBlockValidationPayments(b *testing.B) {
+	benchmarkBlockValidationMix(b, 0)
+}
+
+func benchmarkBlockValidationMix(b *testing.B, numAssets int) {
+	l0, l1, creator, accts, aIdxs, err :=
+		setupEnv(b.TempDir(), b.Name(), numAssets)
+	require.NoError(b, err)
+
+	numBlocks := b.N
+	cert := agreement.Certificate{}
+	var blocks []bookkeeping.Block
+	var txPerBlock int
+	var numAss, numPay int
+	fmt.Printf("Preparing transactions and adding the blocks (/%d): ", numBlocks)
+	evalTime := float64(0)
+	addBlockTime := float64(0)
+	s2 := time.Now()
+	s3 := time.Now()
+
+	for i := 0; i < numBlocks; i++ {
+		// Construct evaluator for next block
+		prev, err := l0.BlockHdr(basics.Round(i + 2))
+		require.NoError(b, err)
+		newBlk := bookkeeping.MakeBlock(prev)
+		eval, err := l0.StartEvaluator(newBlk.BlockHeader, 5000, 0)
+		require.NoError(b, err)
+
+		// build a payset
+		var j int
+		for a, acct := range accts {
+			j++
+			tx := sendAssetTransaction(uint64(j), uint64(i+1), creator, 1, crypto.Digest{1}, basics.AssetIndex(aIdxs[a]), acct)
+			tx.Note = []byte(fmt.Sprintf("%d,%d,%d", i, j, 0))
+			stxn := transactions.SignedTxn{Txn: tx, Sig: crypto.Signature{1}}
+			err = eval.Transaction(stxn, transactions.ApplyData{})
+			if err != nil {
+				require.NoError(b, err)
+			}
+			numAss++
+		}
+
+		receiver := creator
+		receiver[1] = receiver[2]
+		amt := uint64(100000)
+		for {
+			j++
+			tx := createPaymentTransaction(uint64(j), uint64(i+1), creator, receiver, amt, crypto.Digest{1})
+			amt = 1
+			var stxn transactions.SignedTxn
+			stxn.Txn = tx
+			stxn.Sig = crypto.Signature{1}
+			et := time.Now()
+			err = eval.Transaction(stxn, transactions.ApplyData{})
+			evalTime += time.Since(et).Seconds()
+			// check if block is full
+			if err == ledgercore.ErrNoSpace {
+				txPerBlock += eval.PaySetSize()
+				break
+			} else {
+				require.NoError(b, err)
+			}
+			numPay++
+		}
+		vblk, err := eval.GenerateBlock()
+		require.NoError(b, err)
+		abt := time.Now()
+		err = l0.AddBlock(vblk.Block(), cert)
+		addBlockTime += time.Since(abt).Seconds()
+		require.NoError(b, err)
+		blocks = append(blocks, vblk.Block())
+		if i*10%numBlocks == 0 {
+			fmt.Printf("%d%% (%.1fsec) ", i*100/numBlocks, time.Since(s3).Seconds())
+			s3 = time.Now()
+		}
+	}
+	fmt.Printf("\n%s sec total (eval: %.1fsec  addBlock: %.1fsec)\n", time.Since(s2).String(), evalTime, addBlockTime)
+
+	b.Logf("built %d blocks, each with %d txns: %d assets %d pay", numBlocks, txPerBlock/numBlocks, numAss/numBlocks, numPay/numBlocks)
+
+	// eval + add all the (valid) blocks to the second ledger, measuring it this time
+	vc := verify.GetMockedCache(true)
+	tt := time.Now()
+	b.ResetTimer()
+	for _, blk := range blocks {
```

Comment:
Shant, I think I miss a point, why do we need two ledgers? to ensure a block we add is valid?

Reply:
exactly. This also simplifies the debugging effort during the implementation of the benchmark. Can and will be removed eventually.